DrawerKit/DrawerKit/Internal API/PresentationController.swift

```diff
@@ -59,106 +78,206 @@ private extension PresentationController {
+        if smallerThanOrEqual(positionY, 0) {
+            return .fullyExpanded
+        } else if greaterThanOrEqual(positionY, containerViewH) {
+            return .collapsed
+        } else if equal(positionY, drawerPartialY) {
+            return .partiallyExpanded
+        } else {
+            let posY = min(max(positionY, 0), containerViewH)
+            guard clampToNearest else { return .transitioning(posY) }
+            return drawerState(for: clamped(posY))
+        }
     }
 
     var currentDrawerY: CGFloat {
-        get { return presentedView?.frame.origin.y ?? 0 }
-        set { presentedView?.frame.origin.y = newValue }
+        get {
+            let posY = presentedView?.frame.origin.y ?? 0
+            return min(max(posY, 0), containerViewH)
+        }
+        set {
+            let posY = min(max(newValue, 0), containerViewH)
+            presentedView?.frame.origin.y = posY
+        }
     }
 
     var currentDrawerCornerRadius: CGFloat {
-        get { return presentedView?.layer.cornerRadius ?? 0 }
+        get {
+            let radius = presentedView?.layer.cornerRadius ?? 0
+            return min(max(radius, 0), maximumCornerRadius)
+        }
         set {
-            presentedView?.layer.cornerRadius = newValue
+            let radius = min(max(newValue, 0), maximumCornerRadius)
+            presentedView?.layer.cornerRadius = radius
             if #available(iOS 11.0, *) {
                 presentedView?.layer.maskedCorners =
                     [.layerMinXMinYCorner, .layerMaxXMinYCorner]
             } else {
-                presentedView?.roundCorners([.topLeft, .topRight], radius: newValue)
+                presentedView?.roundCorners([.topLeft, .topRight], radius: radius)
+            }
+        }
+    }
+
+    func cornerRadius(at state: DrawerState) -> CGFloat {
+        guard maximumCornerRadius != 0 && drawerPartialY != 0
+            && drawerPartialY != containerViewH else { return 0 }
+
+        let positionY = drawerPositionY(for: state)
+
+        let fraction: CGFloat
+        if supportsPartialExpansion {
+            if positionY < drawerPartialY {
+                fraction = positionY / drawerPartialY
+            } else {
+                fraction = 1 - (positionY - drawerPartialY) / (containerViewH - drawerPartialY)
             }
+        } else {
+            fraction = 1 - positionY / containerViewH
+        }
+
+        return fraction * maximumCornerRadius
+    }
+
+    func clamped(_ positionY: CGFloat) -> CGFloat {
+        if smallerThanOrEqual(positionY, upperMarkY) {
+            return 0
+        } else if greaterThanOrEqual(positionY, lowerMarkY) {
+            return containerViewH
+        } else if smallerThanOrEqual(positionY, drawerPartialY) {
+            return (supportsPartialExpansion ? drawerPartialY : 0)
+        } else {
+            return (supportsPartialExpansion ? drawerPartialY : containerViewH)
```

Comment:
These pharentesis :S

Reply:
Sorry. Unless more people complain about them, I'm leaving them as is. I find that they improve readability.